src/ApiService/ApiService/onefuzzlib/NodeOperations.cs

```diff
@@ -168,14 +175,104 @@ IOnefuzzContext context
+            return;
+        }
+
+        var outdated = this.SearchOutdated(excludeUpdateScheduled: true);
+        await foreach (var node in outdated) {
+            _logTracer.Info($"node is outdated: {node.MachineId} - node_version:{node.Version} api_version:");
+
+            if (node.Version == "1.0.0") {
+                await ToReimage(node, done: true);
+            } else {
+                await ToReimage(node);
+            }
+        }
+    }
+
+
+    public static string SearchOutdatedQuery(
+    string oneFuzzVersion,
+    Guid? poolId = null,
+    Guid? scalesetId = null,
+    IEnumerable<NodeState>? states = null,
+    PoolName? poolName = null,
+    bool excludeUpdateScheduled = false,
+    int? numResults = null) {
+
+        List<string> queryParts = new();
+
+        if (poolId is not null) {
+            queryParts.Add($"(pool_id eq '{poolId}')");
+        }
+
+        if (poolName is not null) {
+            queryParts.Add($"(pool_name eq '{poolName}')");
+        }
+
+        if (scalesetId is not null) {
+            queryParts.Add($"(scaleset_id eq '{scalesetId}')");
+        }
+
+        if (states is not null) {
+            var q = Query.EqualAnyEnum("state", states);
+            queryParts.Add($"({q})");
+        }
+
+        if (excludeUpdateScheduled) {
+            queryParts.Add($"reimage_requested eq false");
+            queryParts.Add($"delete_requested eq false");
+        }
+
+        //# azure table query always return false when the column does not exist
+        //# We write the query this way to allow us to get the nodes where the
+        //# version is not defined as well as the nodes with a mismatched version
+        var versionQuery = $"not (version eq '{oneFuzzVersion}')";
+        queryParts.Add(versionQuery);
+        return Query.And(queryParts);
+    }
+
+    IAsyncEnumerable<Node> SearchOutdated(
+            Guid? poolId = null,
+            Guid? scalesetId = null,
+            IEnumerable<NodeState>? states = null,
+            PoolName? poolName = null,
+            bool excludeUpdateScheduled = false,
+            int? numResults = null) {
+
+        var query = SearchOutdatedQuery(_context.ServiceConfiguration.OneFuzzVersion, poolId, scalesetId, states, poolName, excludeUpdateScheduled, numResults);
+        if (numResults is null) {
+            return QueryAsync(query);
+        } else {
+            return QueryAsync(query).TakeWhile((_, i) => i < numResults);
+        }
+    }
+
+    public async Async.Task CleanupBusyNodesWithoutWork() {
+        //# There is a potential race condition if multiple `Node.stop_task` calls
+        //# are made concurrently.  By performing this check regularly, any nodes
+        //# that hit this race condition will get cleaned up.
+        var nodes = _context.NodeOperations.SearchStates(states: NodeStateHelper.BusyStates);
+
+        await foreach (var node in nodes) {
+            await StopIfComplete(node, true);
+        }
```

Comment:
We can issue all of these in parallel (there are more examples below as well):
```suggestion
        await Task.WhenAll(nodes.Select(n => StopIfComplete(n, true));
```

Reply:
will address in follow up pr